### ISSUE TITLE
Toggle-able world-wide delayed component removal.

### ISSUE THREAD

**Issue by ghroot** [junkdog/artemis-odb#508]
I don't know if this is by design, but in the code below, is the component supposed to be null?

```
public class MySystem extends IteratingSystem {
  private ComponentMapper<MyComponent> myMapper;

  public MySystem() {
    super(Aspect.all(MyComponent.class));
  }

  @Override
  public void removed(int entity) {
    MyComponent myComponent = myMapper.get(entity)
    // myComponent is null here
  }
}
```

If it is by design, is there a way to get a reference to the component that was removed?

**Comment by ghroot** [comment]
Never mind, I just realised you have to add @DelayedComponentRemoval to be able to access the removed component in removed(). Leaving the issue here for anyone else that might run into the same situation.

**Comment by junkdog** [comment]
Yeah, discoverability of annotations is not the best - especially not the edge-casey ones. I haven't found a good way of reporting it at runtime. for a short time, there was a debug build of artemis; it had some preprocessor-driven debug and assertion code injected - but it didn't feel like a robust solution, and was retired when some of the problem areas were refactored away.

Did you find the wiki page summarizing [all annotations](https://github.com/junkdog/artemis-odb/wiki/Annotations)?

**Comment by ghroot** [comment]
Yes I've read about the annotations, just didn't remember that one. Is there a case for making delayed component removal the default?

Also I've noticed that in removed(int entity) mapper.has() and mapper.get() returns different results. has() would return false (event if annotated with delayed removal), but get() would return the component.

**Comment by junkdog** [comment]
Delayed component removal comes with a perf cost. I don't recall the actual numbers though - it was based on running the add/remove component  [benchmarks](https://github.com/junkdog/entity-system-benchmarks). I'm honestly not sure how much of a difference it makes nowadays, but it's primarily useful when toggling components for entities en masse - especially on android devices (I'm in the habit of temporarily marking newly created entities, to facilitate initial placement/transforms of otherwise static assets etc).

What if it was an option in `WorldConfiguration`:

```java
 WorldConfiguration config = new WorldConfiguration()
     .setSystem(new SomeSystemA())
     .setSystem(new SomeSystemB())
     .setDelayedComponentRemoval(true); // maybe s/set/use/ ?
```

It might make it a bit easier to find, while also providing a way to toggle it on a per world-basis. Defaults to current behavior.

This leaves the marker-type components w/o fields: in my experience, these are the ones most likely candidates for frequent add/remove, but since they don't carry any data, it doesn't really make sense to delay their deletion (that is, apart from consistency). That is something which could be dealt with at a later time though - ie, go with all components treated equally for the time being.

Thoughts?

> Also I've noticed that in removed(int entity) mapper.has() and mapper.get() returns different results. has() would return false (event if annotated with delayed removal), but get() would return the component.

Oops, thanks. I'll create a separate issue for it.

**Comment by voidburn** [comment]
@ghroot One example I'm using extensively for removed(), regardless of when it's called in the world's lifecycle is releasing runtime resources to the memory pools. It's a flag that tells me I can safely recycle all the resources that were used by that component instance.

**Comment by junkdog** [comment]
> I suppose I'm wondering if there is any point in implementing the removed() callback in a system if the component wont be available anyway? It feels like in 99% of the cases you need access to it to do anything meaningful.

Definitely use-cases outside of accessing the components - I made a cursory grep through my code; they roughly appear with the same frequency (last item being somewhat less common):

- Only cares about entity (e.g. removing/adding related components)
- Acting on data in removed component (similar to the above)
- Flagging system state as dirty (renderer sorts its own entities)

Circa 1/15 of the components were annotated with `@DelayedComponentRemoval`, though it varied a bit depending on project.

 When dealing with entities/components, I often let the system/entity subscription hold onto any additional data - which is usually only feasible if the data is read-only and can be created during `inserted()`, especially when considering things like serialization.

> In addition to this, there is also the inconsistency between removing a component from an entity and deleting the entity itself. When a component is removed the mapper will return null (unless marked for delayed removal), and when an entity is deleted all components are available. 

Iirc, it was changed to work that way to access inventory dropped by mobs (alternatively, my memory betrays me), prior to`@DelayedComponentRemoval`. 

Current setup favors least amount of overhead; I tried harmonizing component/entity creation/deletion right before 2.0, but I didn't find anything which worked out nicely w/o impacting perf too much.

re nodes: Not familiar with Ash; I'll take a look. I think the c# port of artemis has something similar, but vanilla artemis did not.
